client/asset/dcr/dcr.go

```diff
@@ -3079,6 +3092,286 @@ func (dcr *ExchangeWallet) EstimateRegistrationTxFee(feeRate uint64) uint64 {
 	return (dexdcr.MsgTxOverhead + dexdcr.P2PKHOutputSize*2 + inputCount*dexdcr.P2PKHInputSize) * feeRate
 }
 
+// MakeBondTx creates a time-locked fidelity bond transaction. The V0
+// transaction has two required outputs:
+//
+// Output 0 is a the time-locked bond output of type P2SH with the provided
+// value. The redeem script looks similar to the refund path of an atomic swap
+// script, but with a pubkey hash:
+//
+//	<locktime[4]> OP_CHECKLOCKTIMEVERIFY OP_DROP OP_DUP OP_HASH160 <pubkeyhash[20]> OP_EQUALVERIFY OP_CHECKSIG
+//
+// The pubkey referenced by the script is provided by the caller.
+//
+// Output 1 is a DEX Account commitment. This is an OP_RETURN output that
+// references the provided account ID.
+//
+//	OP_RETURN <account ID>
+//
+// Having the account ID in the raw allows the txn alone to identify the account
+// without the bond output's redeem script.
+//
+// Output 2 is change, if any.
+//
+// The bond output's redeem script, which is needed to spend the bond output, is
+// returned. The bond output pays to a pubkey script for a wallet address.
+// Bond.RedeemTx is a backup transaction that spends the bond output after
+// lockTime passes, paying to an address for the current underlying wallet; the
+// bond private key (BondPrivKey) should normally be used to author a new
+// transaction paying to a new address instead.
+func (dcr *ExchangeWallet) MakeBondTx(ver uint16, amt uint64, lockTime time.Time, bondKey *secp256k1.PrivateKey, acctID []byte) (*asset.Bond, error) {
+	if ver != 0 {
+		return nil, errors.New("only version 0 bonds supported")
+	}
+	if until := time.Until(lockTime); until >= 365*12*time.Hour /* ~6 months */ {
+		return nil, fmt.Errorf("that lock time is nuts: %v", lockTime)
+	} else if until < 0 {
+		return nil, fmt.Errorf("that lock time is already passed: %v", lockTime)
+	}
+
+	pk := bondKey.PubKey().SerializeCompressed()
+	pkh := stdaddr.Hash160(pk)
+
+	feeRate := dcr.targetFeeRateWithFallback(2, 0)
+
+	baseTx := wire.NewMsgTx()
+	const scriptVersion = 0
+
+	// TL output.
+	lockTimeSec := lockTime.Unix()
+	if lockTimeSec > math.MaxInt32 { // https://en.wikipedia.org/wiki/Year_2038_problem
+		return nil, fmt.Errorf("invalid lock time %v", lockTime)
+	}
```

Comment:
I guess that stackexchange Q&A is referring to the block height integer in the serialized block header, whereas in Script, the variable-length integer encoding has a sign bit.

However, I'm revisiting this now and I think that although a ScriptNum is signed, there is a special case in the evaluation of the CLTV opcode that uses _up to_ a 5-byte ScriptNum when _interpreting_ the lockTime bytes.

dcrd: https://github.com/decred/dcrd/blob/d41cb608bfacf91cd36a28108211e9a0096fc282/txscript/opcode.go#L952-L958
bitcoind: https://github.com/bitcoin/bitcoin/blob/f227e153e80c8c50c30d76e1ac638d7206c7ff61/src/script/interpreter.cpp#L531-L545

I'm not sure if that means it is still a 4-byte lockTime in the Script, but this workaround exits to allow values up to the year 2106, or if it can be up to 5 bytes in the serialized script...

I'm afraid we may actually be doing it wrong by assuming 4 bytes only in `ExtractSwapDetails` (as well as in this PR).

Reply:
Ugh, yeah, when we get the bytes for the lockTime data push from `tokenizer.Data()`, we should be feeding that to `txscript.MakeScriptNum(data, 5)`...
Actually a relieve that it's not a 2038 problem, but we've oversimplified by requiring 4 bytes only rather than up to 5.

Reply:
Been making these annoying CLTV lock time changes.  Will push shortly, but in summary, the redeeming **transaction's** lockTime is a 32-byte unsigned integer, which limits the **CLTV lock time value**.  However, the fact that the `ScriptNum` is a signed type does not actually limit to 2038 because there's an exception that allows a 5-byte `ScriptNum` for the CLTV op code.  So, we use a 4 or 5 byte `ScriptNum`, as required, but we still limit the value to 2^32-1 (sometime in 2106).

So that stackexchange is actually sorta relevant because it limits what CLTV lock time values are usable.

Reply:
The 2038 issue is now resolved in https://github.com/decred/dcrdex/pull/1818/commits/69d2394578b6046a3dcc1532565a712fef4a279f